### ISSUE TITLE
v3 function schemas don't show error for missing fields on returns

### ISSUE THREAD

**Issue by mmahalwy** [colinhacks/zod#341]
Given code:
![image](https://user-images.githubusercontent.com/3749095/111741109-26331380-8843-11eb-9755-d90589af7063.png)

The function returns additional fields that are not on the returns schema. The expected behavior is that TS would throw an error due to `key2` being an additional field on the type. Adding `strict()` does not fix this either

**Comment by colinhacks** [comment]
This is unfortunately an inevitable side-effect of how `implement` is...implemented. It doesn't force your function to conform to the *exact* type signature you've defined. Instead it allows any function that `extends` that type signature. The distinction is subtle and requires understanding of how type inference works in TS. This is necessary to allow type inference of the function's return type if you don't provide a `.returns()` schema:

```ts
z.function().args(z.string()).implement(arg => arg.length);
// inferred type: (arg: string)=>number
```

**Comment by mmahalwy** [comment]
Can there be a `implementWithReturns` function that would match the `returns`?

**Comment by colinhacks** [comment]
I just played around with this ideas and I can't get it to work. Here are the implementations of `implement` and `strictImplement` (my preferred name for `implementWithReturns`:

```ts

  implement = <F extends InnerTypeOfFunction<Args, Returns>>(func: F): F => {
    const validatedFunc = this.parse(func);
    return validatedFunc as any;
  };

  strictImplement = (
    func: InnerTypeOfFunction<Args, Returns>
  ): InnerTypeOfFunction<Args, Returns> => {
    const validatedFunc = this.parse(func);
    return validatedFunc as any;
  };

```

Even using strictImplement, your issue isn't resolved:

```ts
const func = z
  .function()
  .args(z.string())
  .returns(z.object({ name: z.string() }))
  .strictImplement((val) => {
    return { name: val, asdf: 1234 };
  });
```

This compiles just fine:
![Screen Shot 2021-03-23 at 12 31 34 PM](https://user-images.githubusercontent.com/3084745/112199375-b8387480-8bd3-11eb-8a1f-c43d7fcedc30.png)

I'm afraid you're hitting against a fundamental property of structural type systems like TypeScript. Consider this:

```ts
type MyFunc = (arg: string) => { name: string };
const myFunc: MyFunc = (arg) => {
  return { name: arg, extraKey: 1234 };
};
```

This also compiles without error, even with the strictest compilation settings. I don't think this is possible.

**Comment by mmahalwy** [comment]
@colinhacks oh yeah, that's strange. Your example at the bottom should error, but TS allows it :S. Let me do some digging to whether there is an issue on the TS side for this.

**Comment by mmahalwy** [comment]
This does error:
```ts
type CustomType = { name: string };

const myFunc = (arg): CustomType => {
  return { name: arg, extraKey: 1234 };
};
```
![image](https://user-images.githubusercontent.com/3749095/112417502-05583b80-8ce5-11eb-9122-f952a6f9b2e1.png)

**Comment by mmahalwy** [comment]
@colinhacks reporting back here, this issue is getting some 👀 from the TS type: https://github.com/microsoft/TypeScript/issues/43365. 

TLDR; have to explicitly set the return type for this to work ATM
```ts
type CustomType = { name: string };

const myFunc = (arg): CustomType => {
  return { name: arg, extraKey: 1234 };
};
```

A middle ground to work for this lib is to infer and set the return type for the function inside of `implement`. Kinda meh about that :S

**Comment by colinhacks** [comment]
Yeah not ideal. This isn't terrible though:

```ts
const func = z
  .function()
  .args(z.string())
  .returns(z.object({ name: z.string() }));

func.implement((arg): ReturnType<z.infer<typeof func>> => {
  // stuff
});
```

**Comment by mmahalwy** [comment]
Ah interesting. I was thinking of this:

```ts
export const func = z
  .function()
  .args(z.string())
  .returns(TestSchema)
  .implement((arg1): z.infer<typeof TestSchema> => {
    // write code
  })
```

**Comment by colinhacks** [comment]
Yep either one is good. 

Note that if `TestSchema` contains transforms you'll want to use `z.input` instead of `z.infer`.
